### PR TITLE
drainer: support relay log for syncer

### PR DESCRIPTION
**Pull request by djshow832**
### What problem does this PR solve? 
Support relay log in Drainer so that the downstream can always be recovered by relay log. Refer to issue #842 for more details.
This is the first part of the feature. This request only supports writing relay log before syncing to the downstream.

### What is changed and how it works?
Add a package `relay`, it supports writing binlog. When `MysqlSyncer` syncs to the downstream, it writes to relay log first. After `MysqlSyncer` successfully writes to the downstream, it purges unused relay log.

### Check List 

Tests 

 - Unit test

Code changes

 - Has exported function/method change
 - Has exported variable/fields change
 - Has interface methods change

Side effects

 - Possible performance regression
 - Increased code complexity

Related changes

 - Need to update the documentation
 - Need to be included in the release note

### CODE REVIEW

File: drainer/relay/relayer.go [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+	tableInfoGetter translator.TableInfoGetter
+	binlogger       binlogfile.Binlogger
+	// lastFileSuffix is the file suffix of last written relay log file.
+	lastFileSuffix uint64
+}
+
+// NewRelayer creates a relayer.
+func NewRelayer(dir string, maxFileSize int64, tableInfoGetter translator.TableInfoGetter) (Relayer, error) {
+	// If `dir` is empty, it means relay log is disabled.
+	if len(dir) == 0 {
```
Comment (july2993) [inline drainer/relay/relayer.go]: return error and check outside if it's disabled, don't call this `NewRelayer`

File: cmd/drainer/drainer.toml [hunk 2]
```diff
@@ -63,6 +63,11 @@ ignore-schemas = "INFORMATION_SCHEMA,PERFORMANCE_SCHEMA,mysql"
 #
 #replicate-do-db = ["~^b.*","s1"]
 
+# directory of relay logs. Empty string indicates disabling relay log.
```
Comment (july2993) [inline cmd/drainer/drainer.toml]: add more explain like only usable when the downstream is tidb/mysql

File: drainer/relay/relayer.go (hunk 1 above)
Comment (djshow832) [reply drainer/relay/relayer.go]: Addressed.

File: cmd/drainer/drainer.toml (hunk 2 above)
Comment (djshow832) [reply cmd/drainer/drainer.toml]: Addressed.

**Comment by july2993** [comment]
/run-integration-tests tidb=release-3.0 tikv=release-3.0 pd=release-3.0

**Comment by july2993** [comment]
ci will fix in https://github.com/pingcap/tidb-binlog/pull/848 not compatible with 3.0 tidb cause some api..
```
[2019-12-13T06:34:11.952Z] [2019/12/13 14:28:55.177 +08:00] [INFO] [panic.go:679] ["publishBinlogs quit"]

[2019-12-13T06:34:11.952Z] panic: runtime error: invalid memory address or nil pointer dereference

[2019-12-13T06:34:11.952Z] [signal SIGSEGV: segmentation violation code=0x1 addr=0x8 pc=0x11f55f6]

[2019-12-13T06:34:11.952Z] 

[2019-12-13T06:34:11.952Z] goroutine 135 [running]:

[2019-12-13T06:34:11.952Z] github.com/pingcap/tidb/store/tikv/tikvrpc.(*Response).GetRegionError(0x0, 0xc000317180, 0xc000d79860, 0xc000370aa0)

[2019-12-13T06:34:11.952Z] 	/go/pkg/mod/github.com/pingcap/tidb@v1.1.0-beta.0.20191119111543-77faf6466821/store/tikv/tikvrpc/tikvrpc.go:669 +0x26

[2019-12-13T06:34:11.952Z] github.com/pingcap/tidb/store/tikv.(*RegionRequestSender).SendReqCtx(0xc000423078, 0xc000317180, 0xc000370aa0, 0x4, 0x1, 0x2, 0x4a817c800, 0xc000510700, 0xc000905fa0, 0x0, ...)

[2019-12-13T06:34:11.952Z] 	/go/pkg/mod/github.com/pingcap/tidb@v1.1.0-beta.0.20191119111543-77faf6466821/store/tikv/region_request.go:151 +0x291

[2019-12-13T06:34:11.952Z] github.com/pingcap/tidb/store/tikv.(*RegionRequestSender).SendReq(...)

[2019-12-13T06:34:11.952Z] 	/go/pkg/mod/github.com/pingcap/tidb@v1.1.0-beta.0.20191119111543-77faf6466821/store/tikv/region_request.go:75

[2019-12-13T06:34:11.952Z] github.com/pingcap/tidb/store/tikv.(*tikvStore).SendReq(0xc00042c1e0, 0xc000317180, 0xc000370aa0, 0x4, 0x1, 0x2, 0x4a817c800, 0x0, 0x3f2dd56cc84415c8, 0xc000423148)

[2019-12-13T06:34:11.952Z] 	/go/pkg/mod/github.com/pingcap/tidb@v1.1.0-beta.0.20191119111543-77faf6466821/store/tikv/kv.go:383 +0xea

[2019-12-13T06:34:11.952Z] github.com/pingcap/tidb/store/tikv.(*LockResolver).getTxnStatus(0xc000185e00, 0xc000317180, 0x5bbf7c893ac0003, 0xc0010384a0, 0x1b, 0x20, 0x5bbf7c893ac0006, 0x5bbf7c893ac0006, 0x0, 0xc0004231a8, ...)

[2019-12-13T06:34:11.952Z] 	/go/pkg/mod/github.com/pingcap/tidb@v1.1.0-beta.0.20191119111543-77faf6466821/store/tikv/lock_resolver.go:471 +0x214

[2019-12-13T06:34:11.952Z] github.com/pingcap/tidb/store/tikv.(*LockResolver).getTxnStatusFromLock(0xc000185e00, 0xc000317180, 0xc000d797c0, 0x5bbf7c893ac0006, 0x40c416, 0xc000d797c0, 0x50, 0x50, 0x15b4d00)

[2019-12-13T06:34:11.952Z] 	/go/pkg/mod/github.com/pingcap/tidb@v1.1.0-beta.0.20191119111543-77faf6466821/store/tikv/lock_resolver.go:404 +0x11b

[2019-12-13T06:34:11.952Z] github.com/pingcap/tidb/store/tikv.(*LockResolver).ResolveLocks(0xc000185e00, 0xc000317180, 0x5bbf7c893ac0006, 0xc000423648, 0x1, 0x1, 0x4a817c800, 0x0, 0xc000314140, 0xc000d61400, ...)

[2019-12-13T06:34:11.952Z] 	/go/pkg/mod/github.com/pingcap/tidb@v1.1.0-beta.0.20191119111543-77faf6466821/store/tikv/lock_resolver.go:295 +0x184

[2019-12-13T06:34:11.952Z] github.com/pingcap/tidb/store/tikv.(*tikvSnapshot).get(0xc000706bd0, 0xc000317180, 0xc0012c8180, 0x2d, 0x34, 0xc000c3a628, 0x18efc80, 0xc000706d50, 0xc000423888, 0x445b6c)

[2019-12-13T06:34:11.952Z] 	/go/pkg/mod/github.com/pingcap/tidb@v1.1.0-beta.0.20191119111543-77faf6466821/store/tikv/snapshot.go:331 +0x45e

[2019-12-13T06:34:11.952Z] github.com/pingcap/tidb/store/tikv.(*tikvSnapshot).Get(0xc000706bd0, 0x18efc00, 0xc000040118, 0xc0012c8180, 0x2d, 0x34, 0x0, 0x0, 0x0, 0x0, ...)

[2019-12-13T06:34:11.952Z] 	/go/pkg/mod/github.com/pingcap/tidb@v1.1.0-beta.0.20191119111543-77faf6466821/store/tikv/snapshot.go:269 +0x1c1

[2019-12-13T06:34:11.952Z] github.com/pingcap/tidb/structure.(*TxStructure).HGet(0xc00053b500, 0x24af058, 0xd, 0xd, 0xc000c3a620, 0x8, 0x8, 0x7f09173d2d65, 0xc00a87e8e0, 0xa87e8e000510700, ...)

[2019-12-13T06:34:11.952Z] 	/go/pkg/mod/github.com/pingcap/tidb@v1.1.0-beta.0.20191119111543-77faf6466821/structure/hash.go:59 +0xdd

[2019-12-13T06:34:11.952Z] github.com/pingcap/tidb/meta.(*Meta).getHistoryDDLJob(0xc000706cc0, 0x24af058, 0xd, 0xd, 0x195, 0x1319bd4, 0x15496c0, 0xc000706cc0)

[2019-12-13T06:34:11.952Z] 	/go/pkg/mod/github.com/pingcap/tidb@v1.1.0-beta.0.20191119111543-77faf6466821/meta/meta.go:642 +0xa6

[2019-12-13T06:34:11.952Z] github.com/pingcap/tidb/meta.(*Meta).GetHistoryDDLJob(0xc000706cc0, 0x195, 0xc000706cc0, 0x0, 0x0)

[2019-12-13T06:34:11.952Z] 	/go/pkg/mod/github.com/pingcap/tidb@v1.1.0-beta.0.20191119111543-77faf6466821/meta/meta.go:655 +0x96

[2019-12-13T06:34:11.952Z] github.com/pingcap/tidb-binlog/drainer.getDDLJob(0x19084a0, 0xc00042c1e0, 0x195, 0x4f36f3, 0xc000a8c215, 0x16d3569)

[2019-12-13T06:34:11.952Z] 	/home/jenkins/agent/workspace/binlog_ghpr_integration/go/src/github.com/pingcap/tidb-binlog/drainer/util.go:138 +0x60

[2019-12-13T06:34:11.952Z] github.com/pingcap/tidb-binlog/drainer.(*Collector).syncBinlog.func1(0x1b, 0xc0010972c0)

[2019-12-13T06:34:11.952Z] 	/home/jenkins/agent/workspace/binlog_ghpr_integration/go/src/github.com/pingcap/tidb-binlog/drainer/collector.go:268 +0x58

[2019-12-13T06:34:11.952Z] github.com/pingcap/tidb-binlog/pkg/util.RetryOnError(0xa, 0x3b9aca00, 0xc001038460, 0x1b, 0xc000423cf0, 0xc001038460, 0x1b)

[2019-12-13T06:34:11.952Z] 	/home/jenkins/agent/workspace/binlog_ghpr_integration/go/src/github.com/pingcap/tidb-binlog/pkg/util/util.go:159 +0xb5

[2019-12-13T06:34:11.952Z] github.com/pingcap/tidb-binlog/drainer.(*Collector).syncBinlog(0xc000544280, 0xc000de63e0, 0x2, 0x1)

[2019-12-13T06:34:11.952Z] 	/home/jenkins/agent/workspace/binlog_ghpr_integration/go/src/github.com/pingcap/tidb-binlog/drainer/collector.go:266 +0x191

[2019-12-13T06:34:11.952Z] github.com/pingcap/tidb-binlog/drainer.(*Collector).publishBinlogs(0xc000544280, 0x18efc80, 0xc0003953b0)

[2019-12-13T06:34:11.952Z] 	/home/jenkins/agent/workspace/binlog_ghpr_integration/go/src/github.com/pingcap/tidb-binlog/drainer/collector.go:136 +0x198

[2019-12-13T06:34:11.952Z] github.com/pingcap/tidb-binlog/drainer.(*Collector).Start.func1(0xc000544280, 0x18efc80, 0xc0003953b0, 0xc000160020)

[2019-12-13T06:34:11.952Z] 	/home/jenkins/agent/workspace/binlog_ghpr_integration/go/src/github.com/pingcap/tidb-binlog/drainer/collector.go:149 +0x3f

[2019-12-13T06:34:11.952Z] created by github.com/pingcap/tidb-binlog/drainer.(*Collector).Start

[2019-12-13T06:34:11.952Z] 	/home/jenkins/agent/workspace/binlog_ghpr_integration/go/src/github.com/pingcap/tidb-binlog/drainer/collector.go:148 +0xa4

[2019-12-13T06:34:11.964Z] [Pipeline] sh
```

File: drainer/relay/relayer.go [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+}
+
+// Close closes binlogger.
+func (r *relayer) Close() error {
+	err := r.binlogger.Close()
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	return nil
```
Comment (suzaku) [inline drainer/relay/relayer.go]: Maybe let `errors.Trace` do the checking like `return errors.Trace(r.binlogger.Close())`?

File: drainer/relay/relayer.go [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+
+	return pos, nil
+}
+
+// GCBinlog removes unused relay log file.
+func (r *relayer) GCBinlog(pos tb.Pos) {
+	// If the file suffix increases, it means previous files are useless.
+	if pos.Suffix > r.lastFileSuffix {
+		r.binlogger.GC(0, pos)
+		r.lastFileSuffix = pos.Suffix
```
Comment (suzaku) [inline drainer/relay/relayer.go]: It looks like this is the only place `lastFileSuffix` is set. Is it `lastGCFileSuffix` instead?

File: drainer/sync/mysql_test.go [hunk 5]
```diff
@@ -87,3 +89,80 @@ func (s *mysqlSuite) TestMySQLSyncerAvoidBlock(c *check.C) {
...
+		select {
+		case <-syncer.Successes():
+		case <-time.After(time.Second):
+			c.Fatal("mysql syncer hasn't synced item in 1s after some error occurs in loader")
+		}
+	}
+
+	names, err := binlogfile.ReadBinlogNames(dir)
+	c.Assert(err, check.IsNil)
+	c.Assert(len(names), check.Equals, 2)
```
[the remark this answers is not in the document]
Comment (djshow832) [reply drainer/sync/mysql_test.go]: I added comments here.
The `binlogger` will rotate if the last file is full, and create a new file before returning.

File: drainer/sync/mysql.go [hunk 6]
```diff
@@ -38,7 +40,7 @@ type MysqlSyncer struct {
 var createDB = loader.CreateDBWithSQLMode
 
 // NewMysqlSyncer returns a instance of MysqlSyncer
-func NewMysqlSyncer(cfg *DBConfig, tableInfoGetter translator.TableInfoGetter, worker int, batchSize int, queryHistogramVec *prometheus.HistogramVec, sqlMode *string, destDBType string) (*MysqlSyncer, error) {
+func NewMysqlSyncer(cfg *DBConfig, tableInfoGetter translator.TableInfoGetter, worker int, batchSize int, queryHistogramVec *prometheus.HistogramVec, sqlMode *string, destDBType string, relayLogDir string, relayLogSize int64) (*MysqlSyncer, error) {
```
[the remark this answers is not in the document]
Comment (djshow832) [reply drainer/sync/mysql.go]: Addressed.

File: drainer/relay/relayer.go (hunk 3 above)
Comment (djshow832) [reply drainer/relay/relayer.go]: Addressed.

File: drainer/relay/relayer.go (hunk 4 above)
Comment (djshow832) [reply drainer/relay/relayer.go]: I changed it to `nextGCFileSuffix`, because the file has not been removed by now.

**Pull request merged**